Fix EnumTypedef name problem

DennisHeimbigner opened the pull request:
## Description of Changes
re: Issue https://github.com/Unidata/netcdf-java/issues/126

The code for handling enum types in Group.java is
incorrect. When creating a new enum type, it is appropriate to
search only the current group for a conflicting name and this is
what the current code does.  But when an enum typed variable
searches for the matching enum type, it must search not only the
current group but all parent groups and the current code does
not do that.

The fix consists of two similar parts.
1. Modify Group.findEnumeration to take an extra boolean parameter
to control if search is this group only or to search up the group's parents.
2. Modify Group.Builder.findEnumTypedef to act like part 1 but to search
the sequence of parent Group.Builder instances.

As a consequence, this PR modifies a number of other files to
adapt to the modified signatures.

## Note
This same problem appears to affect the opaque type also, but
fixing that may be a bit more difficult because CDM appears to
convert opaque types to byte typess.

## PR Checklist
<!-- This will become an interactive checklist once the PR is opened -->
- [x] Link to any issues that the PR addresses
- [ ] Add labels
- [x] Open as a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/)
       until ready for review
- [x] Make sure GitHub tests pass
- [x] Mark PR as "Ready for Review"

## Addendum 1
It turns out that I missed the error in the code in H5headerNew that
attempts to match an enum typed variable with the proper enumeration
type declaration.

The problem and fix (as described by a comment in H5headerNew)
is a bit of a hack. It should be fixed in H5Object.read().
Unfortunately, information is not being passed down so that
the proper fix can be applied early in the HDF5->CDM translation.
Fixing this would affect a lot of function signatures.

Also modified TestEnumTypedef.java to test two cases:
1. the actual enum type def is in the same group as the variable that uses it.
2. the actual enum type def is in a parent group of the variable that uses it

## Misc. Other Changes
* Suppress printing of _NCProperties to simplify text-based comparison testing.

## Addendum 2 (12/19/2022)
* NetcdfFile.java: convert to use H5iospNew (needed by TestH5iosp.java)
* H5headerNew.java: provide get function for accessing the btree (needed by TestDataBTree.java)
* H5iospNew.java: make getRandomAccessFile() method public (needed by tests)
* CompareNetcdf2.java: Add a constructor to specify if attribute name comparison should ignore case or not. It turns out that some tests require case sensitive name matching. Specifically TestCoordSysCompare.java and TestN3iospCompare.java

## Addendum 3 (2/25/2023)
Make additional changes to fix Jenkins failures.

* Revert Group.attributes()
* Revert the use of H5iospNew in NetcdfFile.
* Add to Group a method that looks for a matching enumeration purely structurally, as opposed to name matching. This corresponds to the same method in Group Builder.
* Rebuild the EnumTypedef search algorithm in H5headerNew and H5header to handle cases in Jenins only enum tests.
* Revert various calls to Group.findEnumeration().
* Add some test case data: ref_anon_enum.h5 and test_enum_type.nc
* Convert TestEnumTypedef to JUNIT.parameter form.
* Revert TestH5iosp.
* The URL used in TestSequence is no longer valid, so change to an equivalent url on the remotetest server.

DennisHeimbigner committed:
ckp

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/master'

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/master'

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/master'

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/master'

DennisHeimbigner committed:
partial-update

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Fix EnumTypedef name problem

## Description of Changes
re: Issue https://github.com/Unidata/netcdf-java/issues/126

The code for handling enum types in Group.java is
incorrect. When creating a new enum type, it is appropriate to
search only the current group for a conflicting name and this is
what the current code does.  But when an enum typed variable
searches for the matching enum type, it must search not only the
current group but all parent groups and the current code does
not do that.

The fix consists of two similar parts.
1. Modify Group.findEnumeration to take an extra boolean parameter
to control if search is this group only or to search up the group's parents.
2. Modify Group.Builder.findEnumTypedef to act like \#1 but to search
the sequence of parent Group.Builder instances.

As a consequence, this PR modifies a number of other files to
adapt to the modified signatures.

## Note
This same problem appears to affect the opaque type also, but
fixing that may be a bit more difficult because CDM appears to
convert opaque types to byte typess.

## PR Checklist
<!-- This will become an interactive checklist once the PR is opened -->
- [x] Link to any issues that the PR addresses
- [ ] Add labels
- [x] Open as a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/)
       until ready for review
- [x] Make sure GitHub tests pass
- [ ] Mark PR as "Ready for Review"

haileyajohnson commented on cdm/core/src/main/java/ucar/nc2/Group.java:
```diff
@@ -376,18 +377,20 @@ public boolean removeAttributeIgnoreCase(String attName) {
   /** Find a Enumeration in this or a parent Group, using its short name. */
   @Nullable
-  public EnumTypedef findEnumeration(String name) {
+  public EnumTypedef findEnumeration(String name, boolean searchup) {
```
Since this is public, we can't change the method signature. We can add a second method signature though.

haileyajohnson commented on cdm/core/src/main/java/ucar/nc2/Group.java:
```diff
@@ -1114,6 +1117,26 @@ public Group.Builder commonParent(Group.Builder other) {
     }
 
+    /** Find a Enumeration in this or a parent Group Builder, using its short name. */
+    public Optional<EnumTypedef> findEnumTypedef(String name, boolean searchup) {
```
Same here, we'll need one with a boolean parameter and one with a default.

DennisHeimbigner replied:
Do you object to overloading the method name or should I use
a different name?

haileyajohnson replied:
Overloading is fine

haileyajohnson commented on cdm/core/src/main/java/ucar/nc2/internal/iosp/hdf5/H5headerNew.java:
```diff
@@ -44,11 +39,11 @@
 import ucar.nc2.internal.iosp.hdf5.H5objects.DataObject;
 import ucar.nc2.internal.iosp.hdf5.H5objects.DataObjectFacade;
+import ucar.nc2.internal.iosp.hdf5.H5objects.Filter;
 import ucar.nc2.internal.iosp.hdf5.H5objects.GlobalHeap;
```
Is this being used?

DennisHeimbigner replied:
Is there a gradle task for checking for unused imports?

haileyajohnson replied:
There may be one, but we don't have it implemented, unfortunately.

haileyajohnson commented on dap4/d4tests/src/test/java/dap4/test/TestDAP4Client.java:
```diff
@@ -0,0 +1,406 @@
+  static final boolean DEBUG = false;
+
+  static final boolean BROKEN = false; // on/off known broken tests
+
```
We can use annotations to turn off known broken tests, we shouldn't need a flag.

DennisHeimbigner replied:
Do you object to overloading the method name or should I use
a different name?

DennisHeimbigner replied:
This will be fixed in a later PR addressing major changes to DAP4

haileyajohnson commented on dap4/d4tests/src/test/java/dap4/test/TestDAP4Client.java:
```diff
@@ -0,0 +1,406 @@
+    }
+  }
+
+  boolean defineAllTestcases(String root, String servlet, String server) {
```
JUnit parameterized?

haileyajohnson replied:
I'm happy to take a pass at converting this test to use parameterized if you'd like

DennisHeimbigner replied:
I would say don't bother with fixing DAP4 tests at the moment.
This PR addresses issues completely independent of DAP4.
I do not think there is any point in trying to add DAP4 fixes to it
since I have another branch that is working on that. In any case,
converting the DAP4  tests to use parameterized tests is going to take
time. I think we will have to live with it for a while.

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge branch 'master' into enumfix.dmh

DennisHeimbigner committed:
1. Remove unused import in H5HeaderNew
2. Add overloaded functions to Group.java to restore
   access to original versions of findEnumTypedef and findEnumeration.

DennisHeimbigner committed:
Force re-test

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge branch 'master' into enumfix.dmh

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
It turns out that I missed the error in the code in H5headerNew that
attempts to match an enum typed variable with the proper enumeration
type declaration.

The problem and fix (as described by a comment in H5headerNew)
is a bit of a hack. It should be fixed in H5Object.read().
Unfortunately, information is not being passed down so that
the proper fix can be applied early in the HDF5->CDM translation.
Fixing this would affect a lot of function signatures.

Also modified TestEnumTypedef.java to test two cases:
1. the actual enum type def is in the same group as the variable that uses it.
2. the actual enum type def is in a parent group of the variable that uses it

## Misc. Other Changes
* Suppress printing of _NCProperties to simplify text-based comparison testing.

DennisHeimbigner committed:
Merge branch 'master' into enumfix.dmh

DennisHeimbigner committed:
## Addendum 2

Sigh! Apparently NetcdfFile.java defaulted to using H5iosp instead
of H5iospNew. This meant that many of my changes were being bypassed.

So, modify NetcdfFile to default to H5iospNew.

DennisHeimbigner committed:
Undo change to NetcdfFile.java'

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge branch 'master' into enumfix.dmh

Commit:
Merge branch 'maint-5.x' into enumfix.dmh

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge branch 'master' into enumfix.dmh

DennisHeimbigner committed:
Merge branch 'enumfix.dmh' of https://github.com/DennisHeimbigner/netcdf-java into enumfix.dmh

DennisHeimbigner commented:
/ready

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge branch 'master' into enumfix.dmh

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
test4

DennisHeimbigner committed:
Merge branch 'master' into enumfix.dmh

DennisHeimbigner committed:
NCProperties fix

DennisHeimbigner committed:
## Additional modifications
* NetcdfFile.java: convert to use H5iospNew (needed by TestH5iosp.java)
* H5headerNew.java: provide get function for accessing the btree (needed by TestDataBTree.java)
* H5iospNew.java: make getRandomAccessFile() method public (needed by tests)
* CompareNetcdf2.java: Add a constructor to specify if attribute name comparison should ignore case or not. It turns out that some tests require case sensitive name matching. Specifically TestCoordSysCompare.java and TestN3iospCompare.java

DennisHeimbigner committed:
Apply Spotless

DennisHeimbigner committed:
remove debugging

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge branch 'master' into enumfix.dmh

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge remote-tracking branch 'Unidata/maint-5.x'

DennisHeimbigner committed:
Merge branch 'master' into enumfix.dmh

Commit:
Merge branch 'maint-5.x' into enumfix.dmh

DennisHeimbigner commented:
I give up. The failing test is apparently referencing the file
> /share/testdata/cdmUnitTest/thredds-test-data/local/thredds-test-data/cdmUnitTest/formats/hdf5/samples

and I have no idea how to access it to see what is going on.

tdrwenski commented:
Hi Dennis, the test data in that folder comes from here: https://github.com/Unidata/thredds-test-data. There is a ton of data which takes very long to download. I can try to send or point you to the relevant subset of the data or you can run the tests on Jenkins against your branch.
